sklearn/utils/validation.py

```diff
@@ -1025,27 +1025,29 @@ def _check_sample_weight(sample_weight, X, dtype=None):
     """
     n_samples = _num_samples(X)
 
-    if dtype is not None and dtype not in [np.float32, np.float64]:
+    if dtype is None and hasattr(sample_weight, "dtype"):
+        dtype = sample_weight.dtype
+
+    if dtype not in [np.float32, np.float64]:
         dtype = np.float64
 
-    if sample_weight is None or isinstance(sample_weight, numbers.Number):
-        if sample_weight is None:
-            sample_weight = np.ones(n_samples, dtype=dtype)
-        else:
-            sample_weight = np.full(n_samples, sample_weight,
-                                    dtype=dtype)
+    if sample_weight is None:
+        sample_weight = np.ones(n_samples, dtype=dtype)
+    elif isinstance(sample_weight, numbers.Number):
+        sample_weight = np.full(n_samples, sample_weight, dtype=dtype)
     else:
-        if dtype is None:
-            dtype = [np.float64, np.float32]
-        sample_weight = check_array(
-                sample_weight, accept_sparse=False,
-                ensure_2d=False, dtype=dtype, order="C"
-        )
-        if sample_weight.ndim != 1:
-            raise ValueError("Sample weights must be 1D array or scalar")
-
-        if sample_weight.shape != (n_samples,):
-            raise ValueError("sample_weight.shape == {}, expected {}!"
+        try:
+            sample_weight = np.array(sample_weight, dtype=dtype)
+        except ValueError as e:
+            e.args = (e.args[0] + ". sample weights must a scalar or "
+                                  "a 1D array numeric types",)
```

Comment:
I'm not very enthusiastic about hacking of error messages in `try... except..`, why do we need this? or the try except clause? Is it related to the array function protocol https://github.com/scikit-learn/scikit-learn/issues/14687 ?

Reply:
The most recent commit doesn't have that `try ... except..` block.

Reply:
you sure? I see it in the diff for this PR and if I make a local checkout of latest changes here.